### PR TITLE
feat: use agent-base

### PR DESCRIPTION
**Pull request by lukekarrys**
This allows a single agent to serve as both the http and https agents.

This PR also refactors some things based on feedback in #57.

### CODE REVIEW

**Comment by lukekarrys** [comment]
@rotu This addresses some of the feedback from #57. It's a larger refactor as well, but it would be great to get feedback on it!

**Comment by rotu** [comment]
Well, I didn't understand the original code and I probably won't understand this, but I'll give it a look and see if anything sticks out!

File: lib/agents.js [hunk 1]
```diff
@@ -1,199 +1,203 @@
...
 
-      this.#options = options
-      this.#timeouts = timeouts
-      this.#proxy = proxy ? { proxies: getProxyType(proxy), proxy: urlify(proxy), noProxy } : null
+    if (proxy) {
+      this.#proxy = {
+        proxy: urlify(proxy),
+        noProxy,
+        Agent: getProxyAgent(proxy),
+      }
```
Comment (rotu) [inline lib/agents.js:31]: This needs some better naming. We have class `Agent` with nested properties `Agent.#proxy.proxy`, `Agent.#proxy.agent`, and `Agent.#proxy.noProxy`. It's not at all clear what these fields should mean.

Comment (rotu) [reply lib/agents.js:31]: And similarly we have the `Agent.proxy` getter and the `Agent.getProxy()` function.

**Comment by rotu** [comment]
If I understand the `connect` function and how Signals work correctly, it could look something like below. I'm assuming that `agent-base` would respect a signal that you pass in. No need for `#timeoutConnection`. Though maybe you want to 

```javascript
  async connect (request, options) {
    // if the connection does not have its own lookup function
    // set, then use the one from our options
    options.lookup ??= this.#options.lookup

    const proxy = this.#getProxy(options)
    const abortController = new AbortController()

    if (this.#timeouts.connection) {
      const timeoutSignal = AbortSignal.timeout(this.#timeouts.connection)
      timeoutSignal.addEventListener('abort', (e) => abortController.abort(e.target.reason))
      options.signal?.addEventListener('abort', (e) => abortController.abort(e.target.reason))
    }

    const connectOptions = {
      ...options,
      signal: abortController.signal,
    }
    connectOptions.keepAlive ??= this.keepAlive
    connectOptions.keepAliveInitialDelay ??= this.keepAliveMsecs
    connectOptions.noDelay ??= this.keepAlive
    connectOptions.timeout ??= this.#timeouts.idle

    let socket
    if (proxy) {
      socket = await proxy.connect(request, connectOptions)
    } else {
      await new Promise(resolve => {
        socket = (options.secureEndpoint ? tls : net).connect(connectOptions, () => resolve())
      })
    }

    if (this.#timeouts.idle) {
      socket.setTimeout(this.#timeouts.idle, () => {
        socket.destroy(new Errors.IdleTimeoutError(options))
      })
    }

    return socket
  }
```

File: lib/agents.js [hunk 2]
```diff
@@ -1,199 +1,203 @@
...
+      return proxyCache.get(cacheKey)
     }
 
-    #setIdleTimeout (socket, options) {
-      if (this.#timeouts.idle) {
-        socket.setTimeout(this.#timeouts.idle, () => {
-          socket.destroy(new Errors.IdleTimeoutError(options))
-        })
-      }
+    let { Agent: ProxyAgent } = this.#proxy
```
Comment (rotu) [inline lib/agents.js:64]: Unpacking syntax is less clear here than `let ProxyAgent = this.#proxy.Agent`

File: lib/util.js [hunk 3]
```diff
@@ -1,84 +1,17 @@
...
-    res = await Promise.race(promises.map((p) => p(ac)))
-    ac.abort()
-  } catch (err) {
-    ac.abort()
-    throw err
-  }
-  return res
-}
-
 const urlify = (url) => typeof url === 'string' ? new URL(url) : url
```
Comment (rotu) [inline lib/util.js:3]: Why even have this helper method? The URL constructor accepts a URL object

File: lib/util.js [hunk 4]
```diff
@@ -1,84 +1,17 @@
...
 
 const appendPort = (host, port) => {
```
Comment (rotu) [inline lib/util.js:5]: You don't need this helper function if you're using URL objects. Just use `url.host`.

File: lib/util.js
Comment (rotu) [inline lib/util.js:1]: YAY! I love when files called `util` are deleted :-)

File: lib/util.js (hunk 4 above)
Comment (lukekarrys) [reply lib/util.js]: This function wasn't using URL objects but the request options. But now that this is using `agent-base` (and maybe the default agents do this too) `port` is always set.

File: lib/agents.js (hunk 1 above)
Comment (lukekarrys) [reply lib/agents.js]: These all got renamed to top level private instance variable on the class.

File: lib/util.js (hunk 3 above)
Comment (lukekarrys) [reply lib/util.js]: Good call, removed this.

**Comment by lukekarrys** [comment]
The `connect` function is tricky because `socket.connect()` doesn't take a signal, only `new Socket()` does which we don't have control over with the proxies.

Originally this function was duplicated with a few changes across all 4 types of proxying we do (including not proxying). So that fact that it does a lot is a side effect of moving to using 3rd party agents and wrapping them all with our own `connect()` to handle the timeouts.

I like the idea of using an `AbortSignal` directly a lot, but since we don't have full control over how sockets are created it might not be possible to do in its current state.

**Comment by rotu** [comment]
> The `connect` function is tricky because `socket.connect()` doesn't take a signal, only `new Socket()` does which we don't have control over with the proxies.
> 
> Originally this function was duplicated with a few changes across all 4 types of proxying we do (including not proxying). So that fact that it does a lot is a side effect of moving to using 3rd party agents and wrapping them all with our own `connect()` to handle the timeouts.
> 
> I like the idea of using an `AbortSignal` directly a lot, but since we don't have full control over how sockets are created it might not be possible to do in its current state.

As far as I can tell, we're not using [`socket.connect`](https://nodejs.org/api/net.html#socketconnect) directly, though. We're using [`net.connect`](https://nodejs.org/api/net.html#netconnect) which [*does* install a signal](https://github.com/nodejs/node/blob/d2c7c367741bdcb6f7f77f55ce95a745f0b29fef/lib/net.js#L490C5-L490C31) when creating the underlying socket. It looks like that signal is used to abort destroy the whole socket even after the connecting is finished, so we'd need to decouple the timeout before the function returns.

As for the agent (which I think doesn't take a signal) I don't see where (if anywhere) we call `agent.destroy`, per [docs](https://nodejs.org/api/http.html#class-httpagent):

> It is good practice, to [destroy()](https://nodejs.org/api/http.html#agentdestroy) an Agent instance when it is no longer in use, because unused sockets consume OS resources.

**Review by wraithgar — APPROVED** [review]

**Pull request merged**